CDS-34265-skip-label-selector-tracking-for-k8s-deployments (#809)
harness/developer-hub: docs/continuous-delivery/cd-advanced/cd-kubernetes-category/skip-harness-label-selector-tracking-on-kubernetes-deployments.md

---
title: Skip Harness Label Selector Tracking on Kubernetes Deployments
description: Prevent Harness from using its default Kubernetes label selector harness.io/track --  stable during canary deployments.
sidebar_position: 9
---

You can prevent Harness from using its default Kubernetes label selector `harness.io/track: stable` during canary deployments. Skipping this label can help when you have existing non-Harness deployed services or conflicts in naming. Currently, this feature is behind the feature flag `SKIP_ADDING_TRACK_LABEL_SELECTOR_IN_ROLLING`. Contact [Harness Support](mailto:support@harness.io) to enable the feature. 

> 

Reviewer: You can prevent Harness from using its default Kubernetes label selector, `harness.io/track: stable` during canary deployments. Skipping this label can help when you have existing non-Harness deployed services or naming conflicts. Currently, this feature is behind the feature flag `SKIP_ADDING_TRACK_LABEL_SELECTOR_IN_ROLLING`.

### Review: Labels in Harness Kubernetes canary phases

Default Harness Kubernetes canary deployments are a two phase process that relies on deployment object labels for tracking. 

> 

Reviewer: Default Harness Kubernetes canary deployments are a two-phase process that relies on deployment object labels for tracking.

1. **Canary Phase:**

> 	1. Harness creates a canary version of the Kubernetes deployment object defined in your service definition **Manifests** section.

Reviewer: Canary phase maybe?

	2. All pods in the canary phase have the label `harness.io/track: canary`. Traffic could be routed to these pods using this label.

> 	3. Once that deployment is verified, the canary delete step deletes it by default.    

Reviewer: I didn't understand: canary delete step deletes what?

	Using this method, Harness provides a canary group as a way to test the new build, run your verification, and then roll out to the following primary group.

> 2. **Primary Phase:**

Reviewer: Harness provides a canary group as a way to test the new build, run your verification, and roll out the canary group to the following primary group using this method.
Is this statement still correct after I rephrase? Can we rephrase this?

> 	1. Runs the actual deployment using a Kubernetes rolling update with the number of pods you specify in the **Manifests** files. For example, `replicas: 3`.

Reviewer: Primary phase

	2. All pods in the primary phase have the label `harness.io/track: stable`.

### Review: Invalid value LabelSelector

Kubernetes labels are immutable. If a specific deployment object already exists in the cluster, you cannot change its labels.

If the deployment object already exists before the Harness deployment, it might conflict with the default Harness Kubernetes canary deployment due to the use of the label `harness.io/track`.

If you are deploying different Harness deployment objects to the same cluster, you might encounter a selector error.

```
The Deployment “harness-example-deployment” is invalid: spec.selector:   
  Invalid value: v1.LabelSelector{MatchLabels:map[string]string{“app”:“harness-example”},   
  MatchExpressions:[]v1.LabelSelectorRequirement{}}: field is immutable
```

Most often, you can delete or rename the deployment object. In some case, you will experience downtime when the object restarts. 

> 

Reviewer: In some cases, you will experience downtime when the object restarts.

As an alternative, you can force Harness to skip the `harness.io/track: stable` label in the canary deployment.

### Skipping the tracking label

Once the relevant feature flag is enabled, a Harness Kubernetes canary deployment works like this:

> * If the deployment object already exists in the cluster without the `harness.io/track: stable`, Harness will not add the `harness.io/track: stable` label to the deployment object.

Reviewer: If the deployment object already exists in the cluster without the `harness.io/track: stable` label, ...

---

> title: Skip Harness Label Selector Tracking on Kubernetes Deployments

Reviewer 2: ```suggestion
title: Skip Harness label selector tracking on Kubernetes deployments
```

Reviewer 2: titles should be in sentence case: https://harness.atlassian.net/wiki/spaces/DOCS/pages/21130772574/Documentation+style+guide

description: Prevent Harness from using its default Kubernetes label selector during canary deployments.
sidebar_position: 9
---

> You can prevent Harness from using its default Kubernetes label selector, `harness.io/track: stable` during canary deployments. Skipping this label can help when you have existing non-Harness deployed services or naming conflicts. Currently, this feature is behind the feature flag `SKIP_ADDING_TRACK_LABEL_SELECTOR_IN_ROLLING`. Contact [Harness Support](mailto:support@harness.io) to enable the feature. 

Reviewer 2: ```suggestion
:::note
Currently, this feature is behind the feature flag `SKIP_ADDING_TRACK_LABEL_SELECTOR_IN_ROLLING`. Contact [Harness Support](mailto:support@harness.io) to enable the feature. 
:::

You can prevent Harness from using its default Kubernetes label selector, `harness.io/track: stable`, during canary deployments. Skipping this label can help when you have existing, non-Harness deployed services or naming conflicts. 
```

Reviewer 2: Feature flag notes go at the top of the topic.

> ### Review: Labels in Harness Kubernetes canary phases

Reviewer 2: ```suggestion
## Labels in Harness Kubernetes canary phases
```

Default Harness Kubernetes canary deployments are a two-phase process that relies on deployment object labels for tracking. 

1. **Canary Phase**
	1. Harness creates a canary version of the Kubernetes deployment object defined in your service definition **Manifests** section.
	2. All pods in the canary phase have the label `harness.io/track: canary`. Traffic could be routed to these pods using this label.
	3. Once this deployment is verified, the canary delete step deletes the deployment by default.    
   

> 	Using this method, Harness provides a canary group as a way to test the new build, run your verification, and roll out the canary group to the following primary group.

Reviewer 2: ```suggestion
	Using this method, Harness provides a canary pod set as a way to test the new build, run your verification.
```

2. **Primary Phase**
	1. Runs the actual deployment using a Kubernetes rolling update with the number of pods you specify in the **Manifests** files. For example, `replicas: 3`.
	2. All pods in the primary phase have the label `harness.io/track: stable`.

> ### Review: Invalid value LabelSelector

Reviewer 2: ```suggestion
## Invalid value LabelSelector
```

Reviewer 2: we don't use `Review` or `Step` or `Option` in headings anymore. We have a lot left from before, but moving forward we don't use them.

Kubernetes labels are immutable. If a specific deployment object already exists in the cluster, you cannot change its labels.

If the deployment object already exists before the Harness deployment, it might conflict with the default Harness Kubernetes canary deployment due to the use of the label `harness.io/track`.

If you are deploying different Harness deployment objects to the same cluster, you might encounter a selector error.

```
The Deployment “harness-example-deployment” is invalid: spec.selector:   
  Invalid value: v1.LabelSelector{MatchLabels:map[string]string{“app”:“harness-example”},   
  MatchExpressions:[]v1.LabelSelectorRequirement{}}: field is immutable
```

Most often, you can delete or rename the deployment object. In some cases, you will experience downtime when the object restarts. 

As an alternative, you can force Harness to skip the `harness.io/track: stable` label in the canary deployment.

> ### Skipping the tracking label

Reviewer 2: ```suggestion
## Skipping the tracking label
```
